File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/push/ServiceId.java
package org.whispersystems.signalservice.api.push;

import com.google.protobuf.ByteString;

import org.signal.libsignal.protocol.SignalProtocolAddress;
import org.whispersystems.signalservice.api.util.UuidUtil;

import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.stream.Collectors;

import javax.annotation.Nullable;

/**
 * A wrapper around a UUID that represents an identifier for an account. Today, that is either an {@link ACI} or a {@link PNI}.
 * However, that doesn't mean every {@link ServiceId} is an <em>instance</em> of one of those classes. In reality, we often
 * do not know which we have. And it shouldn't really matter.
 *
 * The only times you truly know, and the only times you should actually care, is during CDS refreshes or specific inbound messages
 * that link them together.
 */
public class ServiceId {

  public static final ServiceId UNKNOWN = ServiceId.from(UuidUtil.UNKNOWN_UUID);

  protected final UUID uuid;

  protected ServiceId(UUID uuid) {
    this.uuid = uuid;
  }

  public static ServiceId from(UUID uuid) {
    return new ServiceId(uuid);
  }

  public static ServiceId parseOrThrow(String raw) {
    return from(UUID.fromString(raw));
  }

  public static ServiceId parseOrThrow(byte[] raw) {
    return from(UuidUtil.parseOrThrow(raw));
  }

  public static @Nullable ServiceId parseOrNull(String raw) {
    UUID uuid = UuidUtil.parseOrNull(raw);
    return uuid != null ? from(uuid) : null;
  }

  public static ServiceId parseOrNull(byte[] raw) {
    UUID uuid = UuidUtil.parseOrNull(raw);
    return uuid != null ? from(uuid) : null;
  }

  public static ServiceId parseOrUnknown(String raw) {
    ServiceId aci = parseOrNull(raw);
    return aci != null ? aci : UNKNOWN;
  }

  public static ServiceId fromByteString(ByteString bytes) {
    return parseOrThrow(bytes.toByteArray());
  }

  public static ServiceId fromByteStringOrNull(ByteString bytes) {
    UUID uuid = UuidUtil.fromByteStringOrNull(bytes);
    return uuid != null ? from(uuid) : null;
  }

  public static ServiceId fromByteStringOrUnknown(ByteString bytes) {
    ServiceId uuid = fromByteStringOrNull(bytes);
    return uuid != null ? uuid : UNKNOWN;
  }

  public UUID uuid() {
    return uuid;
  }

  public boolean isUnknown() {
    return uuid.equals(UNKNOWN.uuid);
  }

  public boolean isValid() {
    return !isUnknown();
  }

  public SignalProtocolAddress toProtocolAddress(int deviceId) {
    return new SignalProtocolAddress(uuid.toString(), deviceId);
  }

  public ByteString toByteString() {
    return UuidUtil.toByteString(uuid);
  }

  public byte[] toByteArray() {
    return UuidUtil.toByteArray(uuid);
  }

  public static List<ServiceId> filterKnown(Collection<ServiceId> serviceIds) {
    return serviceIds.stream().filter(sid -> !sid.equals(UNKNOWN)).collect(Collectors.toList());
  }

  @Override
  public String toString() {
    return uuid.toString();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof ServiceId)) return false;
    final ServiceId serviceId = (ServiceId) o;
    return Objects.equals(uuid, serviceId.uuid);
  }

  @Override
  public int hashCode() {
    return uuid.hashCode();
  }
}

File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/push/ServiceId.kt
package org.whispersystems.signalservice.api.push

import com.google.protobuf.ByteString
import org.signal.libsignal.protocol.ServiceId.InvalidServiceIdException
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.whispersystems.signalservice.api.util.UuidUtil
import java.lang.IllegalArgumentException
import java.util.UUID
import kotlin.jvm.Throws
import org.signal.libsignal.protocol.ServiceId as LibSignalServiceId
import org.signal.libsignal.protocol.ServiceId.Aci as LibSignalAci
import org.signal.libsignal.protocol.ServiceId.Pni as LibSignalPni

/**
 * A wrapper around a UUID that represents an identifier for an account. Today, that is either an [ACI] or a [PNI].
 * However, that doesn't mean every [ServiceId] is an *instance* of one of those classes. In reality, we often
 * do not know which we have. And it shouldn't really matter.
 *
 * The only times you truly know, and the only times you should actually care, is during CDS refreshes or specific inbound messages
 * that link them together.
 */
sealed class ServiceId(val libSignalServiceId: LibSignalServiceId) {
  companion object {
    @JvmStatic
    fun fromLibSignal(serviceId: LibSignalServiceId): ServiceId {
      return when (serviceId) {
        is LibSignalAci -> ACI(serviceId)
        is LibSignalPni -> PNI(serviceId)
        else -> throw IllegalArgumentException("Unknown libsignal ServiceId type!")
      }
    }

    /** Parses a ServiceId serialized as a string. Returns null if the ServiceId is invalid. */
    @JvmStatic
    fun parseOrNull(raw: String?): ServiceId? {
      if (raw == null) {
        return null
      }

      return try {
        fromLibSignal(LibSignalServiceId.parseFromString(raw))
      } catch (e: IllegalArgumentException) {
        null
      } catch (e: InvalidServiceIdException) {
        null
      }
    }

    /** Parses a ServiceId serialized as a byte array. Returns null if the ServiceId is invalid. */
    @JvmStatic
    fun parseOrNull(raw: ByteArray?): ServiceId? {
      if (raw == null) {
        return null
      }

      return try {
        fromLibSignal(LibSignalServiceId.parseFromBinary(raw))
      } catch (e: IllegalArgumentException) {
        null
      } catch (e: InvalidServiceIdException) {
        null
      }
    }

    /** Parses a ServiceId serialized as a ByteString. Returns null if the ServiceId is invalid. */
    @JvmStatic
    fun parseOrNull(bytes: ByteString): ServiceId? = parseOrNull(bytes.toByteArray())

    /** Parses a ServiceId serialized as a string. Crashes if the ServiceId is invalid. */
    @JvmStatic
    @Throws(IllegalArgumentException::class)
    fun parseOrThrow(raw: String): ServiceId = parseOrNull(raw) ?: throw IllegalArgumentException("Invalid ServiceId!")

    /** Parses a ServiceId serialized as a byte array. Crashes if the ServiceId is invalid. */
    @JvmStatic
    @Throws(IllegalArgumentException::class)
    fun parseOrThrow(raw: ByteArray): ServiceId = parseOrNull(raw) ?: throw IllegalArgumentException("Invalid ServiceId!")

    /** Parses a ServiceId serialized as a ByteString. Crashes if the ServiceId is invalid. */
    @JvmStatic
    @Throws(IllegalArgumentException::class)
    fun parseOrThrow(bytes: ByteString): ServiceId = parseOrThrow(bytes.toByteArray())
  }

  val rawUuid: UUID = libSignalServiceId.rawUUID

  val isUnknown: Boolean = rawUuid == UuidUtil.UNKNOWN_UUID

  val isValid: Boolean = !isUnknown

  fun toProtocolAddress(deviceId: Int): SignalProtocolAddress = SignalProtocolAddress(libSignalServiceId.toServiceIdString(), deviceId)

  fun toByteString(): ByteString = ByteString.copyFrom(libSignalServiceId.toServiceIdBinary())

  fun toByteArray(): ByteArray = libSignalServiceId.toServiceIdBinary()

  fun logString(): String = libSignalServiceId.toLogString()

  /**
   * A serialized string that can be parsed via [parseOrThrow], for instance.
   * Basically ACI's are just normal UUIDs, and PNI's are UUIDs with a `PNI:` prefix.
   */
  override fun toString(): String = libSignalServiceId.toServiceIdString()

  data class ACI(val libSignalAci: LibSignalAci) : ServiceId(libSignalAci) {
    companion object {
      @JvmField
      val UNKNOWN = from(UuidUtil.UNKNOWN_UUID)

      @JvmStatic
      fun from(uuid: UUID): ACI = ACI(LibSignalAci(uuid))

      @JvmStatic
      fun parseOrNull(raw: String?): ACI? = ServiceId.parseOrNull(raw).let { if (it is ACI) it else null }

      @JvmStatic
      fun parseOrNull(raw: ByteArray?): ACI? = ServiceId.parseOrNull(raw).let { if (it is ACI) it else null }

      @JvmStatic
      @Throws(IllegalArgumentException::class)
      fun parseOrThrow(raw: String?): ACI = parseOrNull(raw) ?: throw IllegalArgumentException("Invalid ACI!")

      @JvmStatic
      @Throws(IllegalArgumentException::class)
      fun parseOrThrow(raw: ByteArray?): ACI = parseOrNull(raw) ?: throw IllegalArgumentException("Invalid ACI!")

      @JvmStatic
      @Throws(IllegalArgumentException::class)
      fun parseOrThrow(bytes: ByteString): ACI = parseOrThrow(bytes.toByteArray())

      @JvmStatic
      fun parseOrUnknown(bytes: ByteString?): ACI = UuidUtil.fromByteStringOrNull(bytes)?.let { from(it) } ?: UNKNOWN

      @JvmStatic
      fun parseOrUnknown(raw: String?): ACI = parseOrNull(raw) ?: UNKNOWN
    }

    override fun toString(): String = super.toString()
  }

  data class PNI(val libSignalPni: LibSignalPni) : ServiceId(libSignalPni) {
    companion object {
      @JvmField
      var UNKNOWN = from(UuidUtil.UNKNOWN_UUID)

      @JvmStatic
      fun from(uuid: UUID): PNI = PNI(LibSignalPni(uuid))

      @JvmStatic
      fun parseOrNull(raw: String?): PNI? = ServiceId.parseOrNull(raw).let { if (it is PNI) it else null }

      /** Parses a plain UUID (without the `PNI:` prefix) as a PNI. Be certain that whatever you pass to this is for sure a PNI! */
      @JvmStatic
      fun parseUnPrefixedOrNull(raw: String?): PNI? {
        val uuid = UuidUtil.parseOrNull(raw)
        return if (uuid != null) {
          PNI(LibSignalPni(uuid))
        } else {
          null
        }
      }

      @JvmStatic
      fun parseOrNull(raw: ByteArray?): PNI? = ServiceId.parseOrNull(raw).let { if (it is PNI) it else null }

      @JvmStatic
      @Throws(IllegalArgumentException::class)
      fun parseOrThrow(raw: String?): PNI = parseOrNull(raw) ?: throw IllegalArgumentException("Invalid PNI!")

      @JvmStatic
      @Throws(IllegalArgumentException::class)
      fun parseOrThrow(raw: ByteArray?): PNI = parseOrNull(raw) ?: throw IllegalArgumentException("Invalid PNI!")

      @JvmStatic
      @Throws(IllegalArgumentException::class)
      fun parseOrThrow(bytes: ByteString): PNI = parseOrThrow(bytes.toByteArray())

      /** Parses a plain UUID (without the `PNI:` prefix) as a PNI. Be certain that whatever you pass to this is for sure a PNI! */
      @JvmStatic
      @Throws(IllegalArgumentException::class)
      fun parseUnPrefixedOrThrow(raw: String?): PNI = parseUnPrefixedOrNull(raw) ?: throw IllegalArgumentException("Invalid PNI!")
    }

    override fun toString(): String = super.toString()
  }
}
